lib/loglimit/loglimit.go

```diff
@@ -0,0 +1,205 @@
+	// time window.
+	windows map[string]*windowInfo
+}
+
+// entryInfo contains information about a certain log entry.
+type entryInfo struct {
+	// entry is the logger entry.
+	entry *log.Entry
+	// level is the log level at which this log entry should be logged.
+	level log.Level
+	// message is the message message.
+	message string
+	// time is the at which the log reported.
+	time time.Time
+}
+
+// windowInfo contains the information necessary to deduplicate a log
+// entry reported within a certain time window.
+type windowInfo struct {
+	// entryInfo contains information about the first log entry
+	// reported for this window such that entryInfo.time represents
+	// the beginning of this time window.
+	entryInfo
+	// occurrences are the occurrences of logs entries (that share
+	// the same log substring) within this window.
+	occurrences int
+}
+
+// New creates a new log limiter.
+func New(cfg Config) (*LogLimiter, error) {
+	if err := cfg.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &LogLimiter{
+		Config:  cfg,
+		entryCh: make(chan entryInfo, cfg.ChannelSize),
+		windows: make(map[string]*windowInfo, len(cfg.LogSubstrings)),
+	}, nil
+}
+
+// Log sends a log to the log limiter.
+func (l *LogLimiter) Log(entry *log.Entry, level log.Level, args ...any) {
+	l.entryCh <- entryInfo{
+		entry:   entry,
+		level:   level,
+		message: sprintln(args...),
+		time:    l.Clock.Now(),
+	}
+}
+
+// Run runs the log limiter.
+func (l *LogLimiter) Run(ctx context.Context) {
+	t := l.Clock.NewTicker(timeWindowCleanupInterval)
+	defer t.Stop()
+
+	for {
+		select {
+		case e := <-l.entryCh:
+			l.deduplicate(e)
+		case <-t.Chan():
+			l.cleanup()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// deduplicate logs if they should not be deduplicated.
+// Otherwise, it records log occurrences within a certain time window.
+func (l *LogLimiter) deduplicate(e entryInfo) {
+	// Log right away if the log entry should not be deduplicated.
+	deduplicate, logSubstring := l.shouldDeduplicate(&e)
+	if !deduplicate {
+		e.entry.Logln(e.level, e.message)
+		return
+	}
+
+	// If the log should be deduplicated, check if there's an active
+	// window for it (i.e if it has been reported during the past minute).
+	window, ok := l.windows[logSubstring]
+	if ok {
+		// If the log has already been logged, simply increase the
+		// number of occurrences.
+		window.occurrences++
+	} else {
+		// If this is the first occurrence, save the log entry and log it.
+		l.windows[logSubstring] = &windowInfo{
+			entryInfo:   e,
+			occurrences: 1,
+		}
+		e.entry.Logln(e.level, e.message)
+	}
+}
+
+// cleanup removes time windows that have ended, logging the first log message
+// again together with the number of occurrences (of logs that share the same
+// log substring) during the window.
+func (l *LogLimiter) cleanup() {
+	for logSubstring, e := range l.windows {
+		if l.Clock.Now().After(e.time.Add(timeWindow)) {
```

Comment:
nit: Use a consistent time for the cleanup?

```suggestion
	now := l.Clock.Now()
	for logSubstring, e := range l.windows {
		if now.After(e.time.Add(timeWindow)) {
```

Reply:
Done in a9299e3e388e37a01b24e96b4f5fcc01ce5f3752.